File: Array/31_40/31Solution.java
class Solution {
    List<List<Integer>> result=new ArrayList();
    void backtrack(int[] nums,List<Integer> list,int start){
        result.add(new ArrayList(list));
        for(int i=start;i<nums.length;i++){
            list.add(nums[i]);
            backtrack(nums,list,i+1);
            list.remove(list.size()-1);
        }
    }
    public List<List<Integer>> subsets(int[] nums) {
        backtrack(nums,new ArrayList(),0);
        return result;
    }
}

File: Array/31_40/31Solution.py
def subsets(nums):
        def backtrack(first = 0, curr = []):
            if len(curr) == k:  
                output.append(curr[:])
                return
            for i in range(first, n):
                curr.append(nums[i])
                backtrack(i + 1, curr)
                curr.pop()
        
        output = []
        n = len(nums)
        for k in range(n + 1):
            backtrack()
        return output
nums = [ int(x.strip()) for x in input().split(",")]
print(subsets(nums))